File: snapnotes-backend/src/main/java/com/snapnotes/service/NoteService.java
package com.snapnotes.service;

import com.snapnotes.dto.NoteDTO;
import com.snapnotes.entity.Note;
import com.snapnotes.exception.SnapnotesNoteException;

import java.util.List;

public interface NoteService {
    List<NoteDTO> getAllNotes() throws SnapnotesNoteException;
    NoteDTO getNoteById(Long id) throws SnapnotesNoteException;
    List<NoteDTO> getNotesByCategory(String category) throws SnapnotesNoteException;
    Long addNote(NoteDTO noteDTO) throws SnapnotesNoteException;
    void updateNote(NoteDTO noteDTO,Long id) throws SnapnotesNoteException;
    void deleteNote(Long id) throws SnapnotesNoteException;
}

File: snapnotes-frontend/src/app/services/note.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Note } from '../models/note';
import { Observable } from 'rxjs';


@Injectable({
  providedIn: 'root',
})
export class NoteService {
  private baseUrl = "http://localhost:8080/snapnotes/notes"

  constructor(private http: HttpClient){}

  getAllNotes(): Observable<Note[]>{
    return this.http.get<Note[]>(this.baseUrl)
  }

  getNoteById(id: number): Observable<Note>{
    return this.http.get<Note>(`${this.baseUrl}/${id}`)
  }
  
  getNotesByCategory(category: string): Observable<Note[]>{
    return this.http.get<Note[]>(`${this.baseUrl}/category/${category}`)
  }

  addNote(noteDto: Note):Observable<Note>{
    return this.http.post<Note>(this.baseUrl, noteDto)
  }

  updateNote(id: number, noteDto: Note): Observable<void>{
    return this.http.put<void>(`${this.baseUrl}/${id}`,noteDto)
  }

  deleteNote(id: number): Observable<void>{
    return this.http.delete<void>(`${this.baseUrl}/${id}`)
  }
}
